onnxruntime/core/providers/rocm/rocm_profiler.cc

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#if defined(USE_ROCM) && defined(ENABLE_ROCM_PROFILING)
+
+#include <chrono>
+#include <time.h>
+
+#include "rocm_profiler.h"
+#include "RoctracerLogger.h"
+#include "core/common/profiler_common.h"
+
+#define BSIZE 4096
+
+typedef uint64_t timestamp_t;
+static timestamp_t timespec_to_ns(const timespec& time) {
+  return ((timestamp_t)time.tv_sec * 1000000000) + time.tv_nsec;
+}
+
+namespace onnxruntime {
+
+namespace profiling {
+
+RocmProfiler::RocmProfiler()
+: d(&RoctracerLogger::singleton())
+{
+}
+
+RocmProfiler::~RocmProfiler()
+{
+}
+
+bool RocmProfiler::StartProfiling() 
+{
+    d->clearLogs();
+    d->startLogging();
+    return true;
+}
+
+void RocmProfiler::EndProfiling(TimePoint start_time, Events& events)
+{
+  d->stopLogging();
+
+  std::map<uint64_t, std::vector<EventRecord>> event_map;
+  std::map<uint64_t, kernelRow*> kernelLaunches;   // correlation id -> kernel info
+  std::map<uint64_t, copyRow*> copyLaunches;     // correlation id -> copy info
+    
+  // Generate EventRecords
+  int64_t profiling_start = std::chrono::duration_cast<std::chrono::nanoseconds>(start_time.time_since_epoch()).count();
+
+  // Wrong clock again - all the cool kids are doing it
```

Comment:
What is wrong here?

Reply:
Oh, ORT and Kineto are using the "incorrect" clock for profiling.  CLOCK_REALTIME stays in sync with NTP time by making periodic adjustments, those adjustments can make the clock jump backwards.  This makes events swap order, e.g. a kernel executes before it was queued.  There are comments and code in kineto to detect this case and throw out effected events.

CLOCK_MONOTONIC also tries to stay in sync with NTP time but does so by adjusting the clock rate (compressing or expanding time) but keeping the rate positive, i.e. time only moves forward.  This gives monotonically increasing time stamps even in the presence of clock corrections.  The order of events is maintained.

Libroctracer (sanely) reports kernel executions with CLOCK_MONOTONIC time stamps.  I'm collecting my timestamps for the api calls in that same domain.  I then convert to the whatever the upstream demands.

I'm not sure why CLOCK_REALTIME is so prevalent.  Maybe the cupti timestamps are in that domain.  I've no clue how they get away with that.  Maybe I misunderstand something.  I'm not interested in it enough to sort through the whole affair myself.  I am happy to leave a snide comment though. :)  Thanks for noticing.  It might be worth having someone look into it and making an informed decision.

Reply:
I see, thanks for the thorough explanation! Can you please elaborate the comment a bit with CLOCK_MONOTONIC and CLOCK_REALTIME, replacing "// Wrong clock again - all the cool kids are doing it" : )